datashape/coretypes.py

```diff
@@ -1217,31 +1109,20 @@ def to_numpy(ds):
     if isinstance(ds, DataShape):
         # The datashape dimensions
         for dim in ds[:-1]:
-            if isinstance(dim, IntegerConstant):
-                shape += (dim,)
-            elif isinstance(dim, Fixed):
+            if isinstance(dim, Fixed):
                 shape += (dim.val,)
             elif isinstance(dim, TypeVar):
                 shape += (-1,)
             else:
-                raise NotNumpyCompatible('DataShape dimension %s is not'
-                                         ' NumPy-compatible' % dim)
+                raise TypeError('DataShape dimension %s is not '
+                                'NumPy-compatible' % dim)
 
         # The datashape measure
         msr = ds[-1]
     else:
         msr = ds
 
-    try:
-        dtype = msr.to_numpy_dtype()
-    except AttributeError:
-        raise NotNumpyCompatible('DataShape measure %s is not NumPy-compatible'
-                                 % msr)
-
-    if type(dtype) != np.dtype:
-        raise NotNumpyCompatible('Internal Error: Failed to produce NumPy '
-                                 'dtype')
-    return shape, dtype
+    return shape, msr.to_numpy_dtype()
```

Comment:
should we continue to catch the potential `AttributeError` here to reraise as a `TypeError`?
Maybe something like
```python
try:
    to_numpy_dtype = msr.to_numpy_dtype
except AttributeError:
    raise TypeError('DataShape measure %s is not NumPy-compatible' % msr)

return shape, to_numpy_dtype()
```